### PR TITLE
Functionalize evarmap passing in Cases.ml

### PR DESCRIPTION
**Pull request by ppedrot**
This PR turns the imperative reference-passing style used in `cases.ml` into a pure state-passing.

It is made of many tiny commits that compile independently. They need to be squashed before merging the PR, but for now I leave the separate so that it'll be easier to track down a regression in the CI.

### CODE REVIEW

**Comment by ejgallego** [comment]
Funny you are working on this, I was giving it a try yesterday :)

Wouldn't it make sense to port `pretyping` first?

**Comment by SkySkimmer** [comment]
>It is made of many tiny commits that compile independently. They need to be squashed before merging the PR, but for now I leave the separate so that it'll be easier to track down a regression in the CI.

If they compile separately isn't it fine to keep them? Maybe renamed to `cases.ml remove evdref x/y` but otherwise fine.

**Comment by ppedrot** [comment]
> Wouldn't it make sense to port pretyping first?

Unclear, there is an inversion of control in pattern-matching compilation, where the pretyper is actually called recursively. If anything they should be ported at the same time, but that would be a lot of work.

**Comment by ppedrot** [comment]
CI is fine as bedrock failure is spurious. What should I do then, comment all of the tiny commits or just squash them and be done with it? Keeping small commits would be a pain in case of rebase.

**Review by ejgallego — APPROVED** [review]
Looks great. With regards to the small commits, would this be my PR I would squash them as you don't gain a lot I think, the whole PR makes sense as a single logical change but YMMV.

I'd be curious to see a bench of this change, in the past I got surprising speedups due to eliminating refs.

**Comment by ejgallego** [comment]
Secondary pretyper maintainer is @gares , adding him for review / merge / or suggesting an assignee.

**Comment by ejgallego** [comment]
Bench at https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/520

**Comment by ppedrot** [comment]
Once the bench passes I'll squash so as to merge this quickly.

**Comment by herbelin** [comment]
Note potential conflict with #7300. Don't know if easier to resolve on your side or on my side though.

**Comment by ppedrot** [comment]
@herbelin Judging from the diff, conflict is going to be minimal. If I squash then doing it on my side will be easy, otherwise you should rebase as easily.

@ejgallego If you want to be the assignee, I'll squash the commit and force-push. Feel free to merge whenever you want.

**Comment by ejgallego** [comment]
> @ejgallego If you want to be the assignee, I'll squash the commit and force-push. Feel free to merge whenever you want.

I'll happily takes this as I am indeed familiar with what has to be done here. I'll wait tho for the maintainers (@mattam82 and @gares) to give me their blessing.

**Comment by ejgallego** [comment]
The ci build failed due to the squashing :/ new try https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/523

**Comment by ejgallego** [comment]
Will wait until the bench is done, then merge if the proper codeowners don't manisfest themselves.

**Pull request merged**
